File: src/js/test/sniper-test.js
const sp = require("../src/sniper"), assert = require("assert"), sinon = require("sinon"), expect = require("chai").expect,
  u = require("util");

class Sniper{ constructor(){} close(){} price(price, increment,bidder){} fail(msg){}}

describe("スナイパーとトランスレーターのテストだよ", () => {
  const item = "sniper item-123 3000 ", winning = item + "1098, 1098 WINNING", noMsg = "toAuctionMsssage is None",
    joining = item + "0 0 JOINING", bidding = item + "1000 1098 BIDDING";
  describe("とりあえず二つとも生成してみるよ", () => {
    it("スナイパーからね",() => {
      let s = new sp.Sniper(winning); assert.equal(s.name, "sniper"); assert.equal(s.item, "item-123");
      assert.equal(s.stopPrice, 3000); assert.equal(s.lastPrice, 1098); assert.equal(s.lastBid, 1098);
      assert.equal(s.state, "WINNING"); assert.equal(s.msg, noMsg);
    });
    it("トランスレーターもね！！", () => {
      let t = new sp.Translator("hoge", null);
      assert.equal(t.msg, "hoge"); assert.equal(t.sniper, null);
    });
  });
  describe("それぞれの終了メッセージへの対応をテストするよ",() => {
    const close_msg = "SOLVersion: 1.1; Event: CLOSE;";
    it("Sniper:参加要請中や入札中にオークションが閉まると落札失敗", () => {
      let sniper = new sp.Sniper(joining); sniper.close();
      assert.equal(sniper.msg, noMsg);assert.equal(sniper.state, "LOST");
      sniper = new sp.Sniper(bidding); sniper.close();
      assert.equal(sniper.msg, noMsg);assert.equal(sniper.state, "LOST");
    });
    it("Sniper:一位入札中にオークションが閉まると落札成功", () => {
      let sniper = new sp.Sniper(winning); sniper.close();
      assert.equal(sniper.msg, noMsg);assert.equal(sniper.state, "WON");
    });
    it("Translator:オークションが閉まるとSniperのCloseを呼ぶよ", () => {
      let sniper = new Sniper(), spy = sinon.spy(sniper, "close");
      let trans = new sp.Translator(close_msg, sniper); trans.translate();
      expect(spy.calledOnes);
    });
    it("SniperAndTranslator:参加要請中や入札中にオークションが閉まると落札失敗", () => {
      let sniper = new sp.Sniper(joining);
      let trans = new sp.Translator(close_msg, sniper); trans.translate();
      let [msg, state] = sniper.res().split('\n');
      assert.equal(msg, noMsg);assert.equal(state, "item-123 0 0 LOST");
      sniper = new sp.Sniper(bidding);
      trans = new sp.Translator(close_msg, sniper); trans.translate();
      [msg, state] = sniper.res().split('\n');
      assert.equal(msg, noMsg);assert.equal(state, "item-123 1000 1098 LOST");
    });
    it("SniperAndTranslator:一位入札中にオークションが閉まると落札成功", () => {
      let sniper = new sp.Sniper(winning);
      let trans = new sp.Translator(close_msg, sniper); trans.translate();
      let [msg, state] = sniper.res().split('\n');
      assert.equal(msg, noMsg);assert.equal(state, "item-123 1098 1098 WON");
    });
  });
  describe("今度は価格情報への対応をテストだ", () => {
    const price_msg = "Event: PRICE; CurrentPrice: %d; Increment: %d; Bidder: %s;",
      bid_msg = "SOLVersion: 1.1; Command: BID; Price: %d;";
    it("Sniper:限度額以下の価格情報が来ると買い注文をだして入札中になるよ", () => {
      let sniper = new sp.Sniper(joining); sniper.price(1000, 98, "other bidder");
      assert.equal(sniper.msg, u.format(bid_msg, 1098)); assert.equal(sniper.state, "BIDDING");
    });
    it("Sniper:自分名義の価格情報が来ると一位入札中になって注文はしないよ", () => {
      let sniper = new sp.Sniper(bidding); sniper.price(1098, 98, "sniper");
      assert.equal(sniper.msg, noMsg); assert.equal(sniper.state, "WINNING");
    });
    it("Sniper:限度額を超えたら買い注文は出さないし脱落中になるよ", () => {
      let sniper = new sp.Sniper(winning); sniper.price(2950, 51, "other bidder");
      assert.equal(sniper.msg, noMsg); assert.equal(sniper.state, "LOSING");
    });
    it("Translator:価格情報が来るとSniperのPriceを引数付きで呼ぶよ", () => {
      let sniper = new Sniper(); spy = sinon.spy(sniper, "price");
      let trans = new sp.Translator(u.format(price_msg, 1000, 98, "bidder"), sniper); trans.translate();
      expect(spy.getCall(0).args).to.deep.equal([1000, 98, "bidder"]);
    });
    it("SniperAndTranslator:限度額以下の価格情報が来ると買い注文をだして入札中になるよ", () => {
      let sniper = new sp.Sniper(joining);
      let trans = new sp.Translator(u.format(price_msg, 1000, 98, "other bidder"), sniper);
      trans.translate(); let [msg, state] = sniper.res().split('\n');
      assert.equal(sniper.msg, u.format(bid_msg, 1098)); assert.equal(state, "item-123 1000 1098 BIDDING");
    });
    it("SniperAndTranslator:自分名義の価格情報が来ると一位入札中になって注文はしないよ", () => {
      let sniper = new sp.Sniper(bidding);
      let trans = new sp.Translator(u.format(price_msg, 1098, 28, "sniper"), sniper);
      trans.translate(); let [msg, state] = sniper.res().split('\n');
      assert.equal(sniper.msg, noMsg); assert.equal(state, "item-123 1098 1098 WINNING");
    });
    it("SniperAndTranslator:限度額を超えたら買い注文は出さないし脱落中になるよ", () => {
      let sniper = new sp.Sniper(winning);
      let trans = new sp.Translator(u.format(price_msg, 2800, 201, "other one"), sniper);
      trans.translate(); let [msg, state] = sniper.res().split('\n');
      assert.equal(sniper.msg, noMsg); assert.equal(state, "item-123 2800 1098 LOSING");
    });
  });
  describe("最後は異常系を調べるよ",() => {
    it("Sniper:ヘンテコなメッセージを受け取るとエラー発生ね", () => {
      let sniper = new sp.Sniper(winning); sniper.fail("a broken message");
      assert.equal(sniper.msg.includes("a broken message"), true); assert.equal(sniper.state, "FAILED");
    });
    it("SniperAndTranslator:少しヘンテコなメッセージを受け取ってもエラー発生だ", () => {
      let sniper = new sp.Sniper(winning);
      let trans  = new sp.Translator("Event: PRICE; Increment: 24; Bidder: Same one;", sniper);
      trans.translate(); let [msg, state] = sniper.res().split('\n');
      console.log(msg);
      assert.equal(msg.includes("failed: "), true); assert.equal(state, "item-123 0 0 FAILED");
    });
  });
});

/*
mocha src/js/test/sniper-test.js
*/

File: src/haskell/test/SniperTest.hs
import Test.HUnit
import System.IO
import qualified Data.Map as M
import Text.Regex
import Src.Haskell.Src.Sniper(sniper, translator, job)

sol = "SOLVersion: 1.1; "
close_msg = sol ++ "Event: CLOSE;"
no_msg = "toAuctionMessage is None"
price_other = sol ++ "Event: PRICE; CurrentPrice: 1000; Increment: 98; Bidder: someone else;"
price_me = sol ++ "Event: PRICE; CurrentPrice: 1098; Increment: 32; Bidder: sniper;"
price_over = sol ++ "Event: PRICE; CurrentPrice: 2800; Increment: 201; Bidder: a bidder;"
missing_event = sol ++ "Event: ;"
missing_bidder = sol ++ "Event: PRICE; CurrentPrice: 197; Increment: 98; : some one;"
bid_msg = sol ++ "Command: BID; Price: 1098;"
joining = "sniper item-123 3000 0 0 JOINING"
bidding = "sniper item-123 3000 1000 1098 BIDDING"
winning = "sniper item-123 3000 1098 1098 WINNING"

translatorTest = do
  let close = translator close_msg
      price = translator price_other
      broken = translator "a broken message"
      missingEvent = translator missing_event
      missingBidder = translator missing_bidder
      tests = TestList
        [
          "translatorTest 1" ~: "Close" ~=? show close,
          "translatorTest 2" ~: "Price 1000 98 \"someone else\"" ~=? show price,
          "translatorTest 3" ~: "Failed \"failed: msg " ~=? take 20 (show broken),
          "translatorTest 4" ~: "missing field [Event]" ~=? take 21 (drop 51 (show missingEvent)),
          "translatorTest 5" ~: "missing field [Bidder]" ~=? take 22 (drop 102 (show missingBidder))
        ]
  runTestText(putTextToHandle stderr False) tests

sniperTest = do
  let j = sniper (words joining)
      b = sniper (words bidding)
      w = sniper (words winning)
      tests = TestList
        [
          "sniperJoiningTest" ~: no_msg ++ "\n" ++ "item-123 0 0 JOINING" ~=? show j,
          "sniperBiddingTest" ~: no_msg ++ "\n" ++ "item-123 1000 1098 BIDDING" ~=? show b,
          "sniperWinningTest" ~: no_msg ++ "\n" ++ "item-123 1098 1098 WINNING" ~=? show w
        ]
  runTestText(putTextToHandle stderr False) tests

closeWhenJoiningJobTest = do
  let s = sniper (words joining)
      t = translator close_msg
      (msg:state:_) = lines (job s t)
      tests = TestList
        [
          "closeWhenJoiningMsg" ~: no_msg ~=? msg,
          "closeWhenJoiningState" ~: "item-123 0 0 LOST" ~=? state
        ]
  runTestText(putTextToHandle stderr False) tests

closeWhenBiddingJobTest = do
  let s = sniper (words bidding)
      t = translator close_msg
      (msg:state:_) = lines (job s t)
      tests = TestList
        [
          "closeWhenBiddingMsg" ~: no_msg ~=? msg,
          "closeWhenBiddingState" ~: "item-123 1000 1098 LOST" ~=? state
        ]
  runTestText(putTextToHandle stderr False) tests

closeWhenWinningJobTest = do
  let s = sniper (words winning)
      t = translator close_msg
      (msg:state:_) = lines (job s t)
      tests = TestList
        [
          "closeWhenWinningMsg" ~: no_msg ~=? msg,
          "closeWhenWinningState" ~: "item-123 1098 1098 WON" ~=? state
        ]
  runTestText(putTextToHandle stderr False) tests

sniperBiddingTest = do
  let s = sniper (words joining)
      t = translator price_other
      (msg:state:_) = lines (job s t)
      tests = TestList
        [
          "closeWhenWinningMsg" ~: bid_msg ~=? msg,
          "closeWhenWinningState" ~: "item-123 1000 1098 BIDDING" ~=? state
        ]
  runTestText(putTextToHandle stderr False) tests

sniperWinningTest = do
  let s = sniper (words bidding)
      t = translator price_me
      (msg:state:_) = lines (job s t)
      tests = TestList
        [
          "closeWhenWinningMsg" ~: no_msg ~=? msg,
          "closeWhenWinningState" ~: "item-123 1098 1098 WINNING" ~=? state
        ]
  runTestText(putTextToHandle stderr False) tests

sniperLosingTest = do
  let s = sniper (words winning)
      t = translator price_over
      (msg:state:_) = lines (job s t)
      tests = TestList
        [
          "closeWhenWinningMsg" ~: no_msg ~=? msg,
          "closeWhenWinningState" ~: "item-123 2800 1098 LOSING" ~=? state
        ]
  runTestText(putTextToHandle stderr False) tests

sniperFailedTest = do
  let s = sniper (words winning)
      t = translator "a broken message"
      (msg:state:_) = lines (job s t)
      tests = TestList
        [
          "closeWhenWinningMsg" ~: "failed: msg is [a broken message] missing field [Event]."~=? msg,
          "closeWhenWinningState" ~: "item-123 0 0 FAILED" ~=? state
        ]
  runTestText(putTextToHandle stderr False) tests

main = do
  translatorTest
  sniperTest
  closeWhenJoiningJobTest
  closeWhenBiddingJobTest
  closeWhenWinningJobTest
  sniperBiddingTest
  sniperWinningTest
  sniperLosingTest
  sniperFailedTest


{-
runghc src/haskell/test/SniperTest.hs
-}